### PR TITLE
Prevent greedy dask array calculation when creating an Image layer

### LINKED ISSUE #2888: When viewer.add_image() is called on a very large float array without specifying contrast_limits, the data range calculation overwhelms system resources.
## 🐛 Bug

When viewer.add_image() is called on a very large float array without specifying contrast_limits, the data range calculation overwhelms system resources.

## To Reproduce

The following script lazily generates a very large array. Wrapper function can be used to count the number of array-generating calls.
```
import dask.array as da
from dask import delayed
import numpy as np
import napari
import itertools

def counter(func, counter_dict={}):
    counter_dict[func]=0
    def wrap(*args,**kwargs):
        counter_dict[func] += 1
        print("Number of times {} has been called so far {}".format(func.__name__, counter_dict[func]))
        return func(*args,**kwargs)
    return wrap

@counter
def get_random_array():
    return np.random.randint(256, size = (1080,1920), dtype = np.uint8).astype(np.float64)

sample = get_random_array()
lazy_random_array = delayed(get_random_array)
lazy_arrays = [lazy_random_array() for i in range(2000)]
dask_arrays = [da.from_delayed(lazy_array, shape = sample.shape, dtype = sample.dtype) for lazy_array in lazy_arrays]
dask_stack = da.stack(dask_arrays, axis = 0)

viewer = napari.Viewer()
layer = viewer.add_image(data = dask_stack)

napari.run()
```
The above code results in many hundreds of calls to get_random_array() and overwhelms system memory/CPU.
The likely culprit is the layer_utils calc_data_range() function, which nominally checks for array size before performing the min/max calculation. However, in the case of script above the entire array seems to be evaluated when deriving contrast limits.

## Expected behavior

Correctly prevent evaluation of excessively large arrays for contrast limits check.

## Environment

napari: 0.4.9
Platform: Windows-10-10.0.19041-SP0
Python: 3.8.10 | packaged by conda-forge | (default, May 11 2021, 06:25:23) [MSC v.1916 64 bit (AMD64)]
Qt: 5.12.9
PyQt5: 5.12.3
NumPy: 1.20.2
SciPy: 1.6.3
Dask: 2021.05.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 NVIDIA 462.31
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 2560x1440, scale 1.0
- screen 2: resolution 1440x2560, scale 1.0

Plugins:
- console: 0.0.3
- scikit-image: 0.4.9.dev8+g43a604be
- svg: 0.1.5

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR makes sure that we don't greedily compute a dask array.  Together with #3022, this fixes #2888  (I think? @jni?)

would be nice to get this on for 0.4.11

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?
test added

### CODE REVIEW

File: napari/layers/image/image.py [hunk 1]
```diff
@@ -292,7 +294,13 @@ def __init__(
         )
         self._experimental_clipping_planes = ClippingPlaneList()
         if contrast_limits is None:
-            self.contrast_limits_range = self._calc_data_range()
+            if isinstance(data, da.Array):
+                if np.issubdtype(data.dtype, np.integer):
```
Comment (tlambert03) [inline napari/layers/image/image.py]: alternative is `not isinstance(obj, np.ndarray)`?

Comment (sofroniewn) [reply napari/layers/image/image.py]: I think that might be better - probably nice to to special case dask here over zarr say. I think reasonable to just do calculation for numpy arrays.

Comment (tlambert03) [reply napari/layers/image/image.py]: changed

**Review by sofroniewn — APPROVED** [review]
Nice - ok I like this now! I'm going to merge now as I'd like to get this in for the next RC, we can change during the rc process if @jni really doesn't like - but i think this is solution he wanted too

**Pull request merged**